src/bridge/BridgeTransport.cpp

```diff
@@ -0,0 +1,138 @@
+}
+
+void BridgeTransport::Stop() {
+    if (!thread_ || !thread_->joinable()) return;
+    StopAsync();
+    logger_->Log("stopping");
+    thread_->join();
+    thread_.reset();
+}
+
+void BridgeTransport::StopAsync() {
+    if (!stop_signal_handle_ || stop_signal_handle_->closing()) return;
+    stop_signal_handle_->send();
+}
+
+void BridgeTransport::RunThread() {
+    logger_->Log("thread started");
+    loop_ = uvw::loop::create();
+    stop_signal_handle_ = GetLoop()->resource<uvw::async_handle>();
+    write_signal_handle_ = GetLoop()->resource<uvw::async_handle>();
+
+    stop_signal_handle_->on<uvw::async_event>([this](const uvw::async_event&, uvw::async_handle& handle) {
+        logger_->Log("closing handles");
+        CloseConnectionHandles();
+        write_signal_handle_->close();
+        stop_signal_handle_->close();
+    });
+    
+    write_signal_handle_->on<uvw::async_event>([this](const uvw::async_event&, uvw::async_handle& handle) {
+        SendWrites();
+    });
+    
+    CreateConnection();
+    GetLoop()->run();
+    GetLoop()->close();
+    logger_->Log("thread exited");
+}
+
+void BridgeTransport::ResetBuffers() {
+    recv_buf_.Clear();
+    send_buf_.Clear();
+}
+
+void BridgeTransport::OnRecv(const uvw::data_event& event) {
+    if (!recv_buf_.Push(event.data.get(), event.length)) {
+        logger_->Log("recv_buf_.Push(%i) failed", event.length);
+        ResetConnection();
+        return;
+    }
+
+    size_t available;
+    while (available = recv_buf_.BytesAvailable()) {
+        if (available < 4) return;
+
+        char len_buf[4];
+        recv_buf_.Peek(len_buf, 4);
+        uint32_t size = LE32_TO_NATIVE(*reinterpret_cast<uint32_t*>(len_buf));
+
+        if (size > VRBRIDGE_MAX_MESSAGE_SIZE) {
+            logger_->Log("message size overflow");
+            ResetConnection();
+            return;
+        }
+
+        auto unwrapped_size = size - 4;
+        if (available < unwrapped_size) return;
+
+        auto message_buf = std::make_unique<char[]>(size);
+        if (!recv_buf_.Skip(4) || !recv_buf_.Pop(message_buf.get(), unwrapped_size)) {
+            logger_->Log("recv_buf_.Pop(%i) failed", size);
+            ResetConnection();
+            return;
+        }
+
+        messages::ProtobufMessage message;
+        if (message.ParseFromArray(message_buf.get(), unwrapped_size)) {
+            message_callback_(message);
+        } else {
+            logger_->Log("receivedMessage.ParseFromArray failed");
+            ResetConnection();
+            return;
+        }
+    }
+}
+
+void BridgeTransport::SendBridgeMessage(const messages::ProtobufMessage& message) {
+    if (!IsConnected()) return;
+
+    uint32_t size = static_cast<uint32_t>(message.ByteSizeLong());
+    uint32_t wrapped_size = size + 4;
+    
+    auto message_buf = std::make_unique<char[]>(wrapped_size);
+    *reinterpret_cast<uint32_t*>(message_buf.get()) = NATIVE_TO_LE32(wrapped_size);
```

Comment:
nit: I'd rather you write the bytes individually over defining a byteswap macro. Or write a macro that does the same.

```suggestion
    message_buf.get()[0] = (wrapped_size >>  0) & 0xFF;
    message_buf.get()[1] = (wrapped_size >>  8) & 0xFF;
    message_buf.get()[2] = (wrapped_size >> 16) & 0xFF;
    message_buf.get()[4] = (wrapped_size >> 24) & 0xFF;
```
This will always work, regardless of endian. You can wrap that in a macro if you prefer, as `WRITE_LE32` or similar.